File: butler_main/butler_flow/surface/dto.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass(frozen=True, slots=True)
class FlowSummaryDTO:
    flow_id: str = ""
    workflow_kind: str = ""
    effective_status: str = ""
    effective_phase: str = ""
    attempt_count: int = 0
    max_attempts: int = 0
    max_phase_attempts: int = 0
    max_runtime_seconds: int = 0
    runtime_elapsed_seconds: int = 0
    goal: str = ""
    guard_condition: str = ""
    approval_state: str = "not_required"
    execution_mode: str = ""
    session_strategy: str = ""
    active_role_id: str = ""
    role_pack_id: str = ""
    last_judge: str = ""
    latest_judge_decision: dict[str, Any] = field(default_factory=dict)
    last_operator_action: str = ""
    latest_operator_action: dict[str, Any] = field(default_factory=dict)
    queued_operator_updates: list[Any] = field(default_factory=list)
    latest_token_usage: dict[str, Any] = field(default_factory=dict)
    context_governor: dict[str, Any] = field(default_factory=dict)
    latest_handoff_summary: dict[str, Any] = field(default_factory=dict)
    updated_at: str = ""

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class RoleRuntimeDTO:
    active_role_id: str = ""
    role_sessions: dict[str, Any] = field(default_factory=dict)
    pending_handoffs: list[dict[str, Any]] = field(default_factory=list)
    recent_handoffs: list[dict[str, Any]] = field(default_factory=list)
    latest_handoff_summary: dict[str, Any] = field(default_factory=dict)
    latest_role_handoffs: dict[str, Any] = field(default_factory=dict)
    role_chips: list[dict[str, Any]] = field(default_factory=list)
    roles: list[dict[str, Any]] = field(default_factory=list)
    execution_mode: str = ""
    session_strategy: str = ""
    role_pack_id: str = ""

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class SupervisorViewDTO:
    header: dict[str, Any] = field(default_factory=dict)
    events: list[dict[str, Any]] = field(default_factory=list)
    latest_supervisor_decision: dict[str, Any] = field(default_factory=dict)
    latest_judge_decision: dict[str, Any] = field(default_factory=dict)
    latest_operator_action: dict[str, Any] = field(default_factory=dict)
    latest_handoff_summary: dict[str, Any] = field(default_factory=dict)
    context_governor: dict[str, Any] = field(default_factory=dict)
    latest_token_usage: dict[str, Any] = field(default_factory=dict)
    pointers: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class WorkflowViewDTO:
    events: list[dict[str, Any]] = field(default_factory=list)
    runtime_summary: dict[str, Any] = field(default_factory=dict)
    artifact_refs: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class ManageCenterDTO:
    preflight: dict[str, Any] = field(default_factory=dict)
    assets: dict[str, Any] = field(default_factory=dict)
    selected_asset: dict[str, Any] = field(default_factory=dict)
    role_guidance: dict[str, Any] = field(default_factory=dict)
    review_checklist: list[str] = field(default_factory=list)
    bundle_manifest: dict[str, Any] = field(default_factory=dict)
    manager_notes: str = ""

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class FlowDetailDTO:
    summary: dict[str, Any] = field(default_factory=dict)
    step_history: list[dict[str, Any]] = field(default_factory=list)
    timeline: list[dict[str, Any]] = field(default_factory=list)
    turns: list[dict[str, Any]] = field(default_factory=list)
    actions: list[dict[str, Any]] = field(default_factory=list)
    artifacts: list[dict[str, Any]] = field(default_factory=list)
    handoffs: list[dict[str, Any]] = field(default_factory=list)
    flow_definition: dict[str, Any] = field(default_factory=dict)
    runtime_snapshot: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

File: butler_main/products/butler_flow/desktop/src/shared/dto.ts
export type FlowStatus = "running" | "paused" | "completed" | "failed" | "draft" | string;

export interface SurfaceMetaDTO {
  surface_id?: string;
  canonical_surface?: string;
  projection_kind?: string;
  display_title?: string;
  legacy_aliases?: string[];
  truth_basis?: string[];
  title?: string;
}

export interface FlowSummaryDTO {
  flow_id: string;
  workflow_kind: string;
  effective_status: FlowStatus;
  effective_phase: string;
  attempt_count: number;
  max_attempts: number;
  max_phase_attempts: number;
  max_runtime_seconds: number;
  runtime_elapsed_seconds: number;
  goal: string;
  guard_condition: string;
  approval_state: string;
  execution_mode: string;
  session_strategy: string;
  active_role_id: string;
  role_pack_id: string;
  last_judge: string;
  latest_judge_decision: Record<string, unknown>;
  last_operator_action: string;
  latest_operator_action: Record<string, unknown>;
  queued_operator_updates: unknown[];
  latest_token_usage: Record<string, unknown>;
  context_governor: Record<string, unknown>;
  latest_handoff_summary: Record<string, unknown>;
  updated_at: string;
}

export interface TimelineEvent {
  event_id: string;
  kind: string;
  flow_id: string;
  phase: string;
  attempt_no: number;
  created_at: string;
  message: string;
  title?: string;
  lane?: string;
  family?: string;
  raw_text?: string;
  payload?: Record<string, unknown>;
}

export interface FlowDetailDTO {
  summary: FlowSummaryDTO;
  step_history: Array<Record<string, unknown>>;
  timeline: TimelineEvent[];
  turns: Array<Record<string, unknown>>;
  actions: Array<Record<string, unknown>>;
  artifacts: Array<Record<string, unknown>>;
  handoffs: Array<Record<string, unknown>>;
  flow_definition: Record<string, unknown>;
  runtime_snapshot: Record<string, unknown>;
}

export interface SupervisorViewDTO {
  header: Record<string, unknown>;
  events: TimelineEvent[];
  latest_supervisor_decision: Record<string, unknown>;
  latest_judge_decision: Record<string, unknown>;
  latest_operator_action: Record<string, unknown>;
  latest_handoff_summary: Record<string, unknown>;
  context_governor: Record<string, unknown>;
  latest_token_usage: Record<string, unknown>;
  pointers: Record<string, unknown>;
}

export interface WorkflowViewDTO {
  events: TimelineEvent[];
  runtime_summary: Record<string, unknown>;
  artifact_refs: string[];
}

export interface RoleRuntimeDTO {
  active_role_id: string;
  role_sessions: Record<string, unknown>;
  pending_handoffs: Array<Record<string, unknown>>;
  recent_handoffs: Array<Record<string, unknown>>;
  latest_handoff_summary: Record<string, unknown>;
  latest_role_handoffs: Record<string, unknown>;
  role_chips: Array<Record<string, unknown>>;
  roles: Array<Record<string, unknown>>;
  execution_mode: string;
  session_strategy: string;
  role_pack_id: string;
}

export interface ManageCenterDTO {
  surface_meta?: SurfaceMetaDTO;
  preflight: Record<string, unknown>;
  assets: {
    items?: Array<Record<string, unknown>>;
    [key: string]: unknown;
  };
  selected_asset: Record<string, unknown>;
  contract_studio?: Record<string, unknown>;
  role_guidance: Record<string, unknown>;
  review_checklist: string[];
  bundle_manifest: Record<string, unknown>;
  manager_notes: string;
}

export interface WorkspacePayload {
  surface_meta?: SurfaceMetaDTO;
  preflight: Record<string, unknown>;
  flows: {
    items: Array<Record<string, unknown>>;
    [key: string]: unknown;
  };
}

export interface SingleFlowPayload extends FlowDetailDTO {
  flow_id: string;
  surface_meta?: SurfaceMetaDTO;
  status: Record<string, unknown>;
  task_contract?: Record<string, unknown>;
  task_contract_summary?: Record<string, unknown>;
  governance_summary?: Record<string, unknown>;
  mission_console?: Record<string, unknown>;
  latest_receipt_summary?: Record<string, unknown>;
  latest_governance_receipt_summary?: Record<string, unknown>;
  latest_artifact_ref?: string;
  accepted_receipt_count?: number;
  recovery_cursor?: Record<string, unknown>;
  recovery_state?: string;
  navigator_summary: FlowSummaryDTO;
  supervisor_view: SupervisorViewDTO;
  workflow_view: WorkflowViewDTO;
  inspector: Record<string, unknown>;
  role_strip: RoleRuntimeDTO;
  operator_rail: Record<string, unknown>;
  flow_console: Record<string, unknown>;
  surface: Record<string, unknown>;
}

export interface DesktopActionPayload {
  configPath?: string;
  flowId: string;
  type: string;
  instruction?: string;
  repoContractPath?: string;
}

export interface DesktopBridgeResult {
  ok: boolean;
  error_type?: string;
  message?: string;
}
